wgpu-native/src/track.rs

```diff
@@ -112,6 +112,45 @@ pub enum Stitch {
     Last,
 }
 
+pub struct ConsumeIterator<'a, I: TypedId, U: Copy + PartialEq> {
+    src: Iter<'a, Index, Track<U>>,
+    dst: &'a mut FastHashMap<Index, Track<U>>,
+    stitch: Stitch,
+    _marker: PhantomData<I>,
+}
+
+impl<'a, I: TypedId, U: Copy + PartialEq> Iterator for ConsumeIterator<'a, I, U> {
+    type Item = (I, Range<U>);
+    fn next(&mut self) -> Option<Self::Item> {
+        loop {
+            let (&index, new) = self.src.next()?;
+            match self.dst.entry(index) {
+                Entry::Vacant(e) => {
+                    e.insert(new.clone());
+                }
+                Entry::Occupied(mut e) => {
+                    assert_eq!(e.get().epoch, new.epoch);
+                    let old = mem::replace(&mut e.get_mut().last, new.last);
+                    if old != new.init {
+                        let state = match self.stitch {
+                            Stitch::Init => new.init,
+                            Stitch::Last => new.last,
+                        };
+                        return Some((I::new(index, new.epoch), old .. state))
+                    }
+                }
+            }
+        }
+    }
+}
+
+// Make sure to finish all side effects on drop
+impl<'a, I: TypedId, U: Copy + PartialEq> Drop for ConsumeIterator<'a, I, U> {
+    fn drop(&mut self) {
+        while self.next().is_some() {}
```

Comment:
Canonical way to do this is `.for_each(drop)`

Reply:
oh wow, fancy :)

Reply:
References:
https://github.com/rust-lang/rust/commit/5c58eec0bd8cee8fb2a191396d5ad5b5c9b0116a
https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.collect

Reply:
Let's switch to `.for_each(drop)` here instead